File: sdks/javascript/src/revocations.ts
/**
 * Revocation operations.
 */

import type { ATAPClient } from "./client.js";
import type { Revocation, RevocationList } from "./models.js";

/** Submit and query revocations. */
export class RevocationAPI {
  private readonly _client: ATAPClient;

  constructor(client: ATAPClient) {
    this._client = client;
  }

  /**
   * Submit a revocation. Requires atap:revoke scope.
   *
   * @param approvalId - The approval ID to revoke (apr_...).
   * @param signature - JWS signature.
   * @param options.validUntil - Optional RFC3339 expiry (defaults to revoked_at + 60min).
   */
  async submit(
    approvalId: string,
    signature: string,
    options: { validUntil?: string } = {},
  ): Promise<Revocation> {
    const body: Record<string, unknown> = {
      approval_id: approvalId,
      signature,
    };
    if (options.validUntil) body.valid_until = options.validUntil;

    const data = await this._client._authedRequest(
      "POST",
      "/v1/revocations",
      { jsonBody: body },
    );
    return parseRevocation(data);
  }

  /**
   * Query active revocations for an entity (public endpoint).
   *
   * @param entityDid - The approver DID to query.
   */
  async list(entityDid: string): Promise<RevocationList> {
    const data = await this._client._http.request("GET", "/v1/revocations", {
      params: { entity: entityDid },
    });
    const revocations = (
      (data.revocations as Record<string, unknown>[]) || []
    ).map(parseRevocation);
    return {
      entity: (data.entity as string) || entityDid,
      revocations,
      checkedAt: (data.checked_at as string) || "",
    };
  }
}

export function parseRevocation(data: Record<string, unknown>): Revocation {
  return {
    id: (data.id as string) || "",
    approvalId: (data.approval_id as string) || "",
    approverDid: (data.approver_did as string) || "",
    revokedAt: (data.revoked_at as string) || "",
    expiresAt: (data.expires_at as string) || "",
  };
}

File: sdks/go/revocations.go
package atap

import "context"

// RevocationAPI provides revocation submission and querying.
type RevocationAPI struct {
	client *Client
}

// Submit submits a revocation. Requires atap:revoke scope.
func (r *RevocationAPI) Submit(ctx context.Context, approvalID, signature string, validUntil string) (*Revocation, error) {
	body := map[string]interface{}{
		"approval_id": approvalID,
		"signature":   signature,
	}
	if validUntil != "" {
		body["valid_until"] = validUntil
	}

	data, err := r.client.authedRequest(ctx, "POST", "/v1/revocations", &RequestOptions{JSONBody: body})
	if err != nil {
		return nil, err
	}
	return parseRevocation(data), nil
}

// List queries active revocations for an entity (public endpoint).
func (r *RevocationAPI) List(ctx context.Context, entityDID string) (*RevocationList, error) {
	data, err := r.client.http.Request(ctx, "GET", "/v1/revocations", &RequestOptions{
		Params: map[string]string{"entity": entityDID},
	})
	if err != nil {
		return nil, err
	}

	var revocations []Revocation
	if items, ok := data["revocations"].([]interface{}); ok {
		for _, item := range items {
			if m, ok := item.(map[string]interface{}); ok {
				revocations = append(revocations, *parseRevocation(m))
			}
		}
	}

	return &RevocationList{
		Entity:      getString(data, "entity"),
		Revocations: revocations,
		CheckedAt:   getString(data, "checked_at"),
	}, nil
}

func parseRevocation(data map[string]interface{}) *Revocation {
	return &Revocation{
		ID:          getString(data, "id"),
		ApprovalID:  getString(data, "approval_id"),
		ApproverDID: getString(data, "approver_did"),
		RevokedAt:   getString(data, "revoked_at"),
		ExpiresAt:   getString(data, "expires_at"),
	}
}
